sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp

```diff
@@ -77,16 +100,30 @@ template <typename Compare = std::less<>> class default_sorter {
   template <typename Group, typename T>
   T operator()([[maybe_unused]] Group g, T val) {
 #ifdef __SYCL_DEVICE_ONLY__
+    // Adjust the scratch pointer based on alignment of the type T.
     // Per extension specification if scratch size is less than the value
     // returned by memory_required then behavior is undefined, so we don't check
     // that the scratch size statisfies the requirement.
-    auto range_size = g.get_local_range().size();
-    size_t local_id = g.get_local_linear_id();
-    T *temp = reinterpret_cast<T *>(scratch.data());
-    ::new (temp + local_id) T(val);
-    sycl::detail::merge_sort(g, temp, range_size, comp,
-                             scratch.data() + range_size * sizeof(T));
-    val = temp[local_id];
+    void *scratch_ptr = scratch.data();
+    size_t space = scratch.size();
+    const size_t range_size = g.get_local_range().size();
+    const size_t local_id = g.get_local_linear_id();
+    scratch_ptr =
+        std::align(alignof(T), /* output storage and temporary storage */ 2 *
+                                   range_size * sizeof(T),
+                   scratch_ptr, space);
+    std::byte *aligned_scratch_ptr = static_cast<std::byte *>(scratch_ptr);
+    // Each work item processes two locations: in the output buffer and in the
+    // temp buffer.
+    T *local_copy = ::new (aligned_scratch_ptr + local_id * sizeof(T)) T(val);
+    ::new (aligned_scratch_ptr + range_size * sizeof(T) + local_id * sizeof(T))
+        T(val);
+    // Broadcast leader's pointer (the beginning of the scratch) to all work
+    // items in the group.
+    T *scratch_begin = sycl::group_broadcast(g, local_copy);
```

Comment:
Sorry for stupid question. Can we replace `group_broadcast` with more lightweight solution?

Reply:
I think that is actually a very complicated question... https://en.cppreference.com/w/cpp/utility/launder seems to be related but I don't know how it plays with memory model/WIs synchronization rules/etc. In other words, would it be legal to use it, if some non-leader WI does `std::launder` before leader had a chance to perform placement new?

\+ @gmlueck , if he has some thoughts about it.

Reply:
I didn't look at the code enough to understand how the algorithm works.  However, the `group_broadcast` operation also acts as a barrier, waiting for all work-items to arrive before any work-item progresses beyond the call to `group_barrier`.  Calling `std::launder` wouldn't have this effect.  Does the algorithm expect a barrier operation here?